File: haskell/glue/src/Glue/Lib/Builtin/Try.hs
module Glue.Lib.Builtin.Try where

import Data.Maybe (maybeToList)
import Data.Text (Text)
import Glue.Eval (Eval, apply, eval, getRuntime, isCallable, liftIO, putRuntime, runEval, throwError)
import Glue.Eval.Error (EvalError (..))
import Glue.Eval.Exception
import Glue.IR qualified as IR

tryFunc :: IR.IR Eval
tryFunc = IR.Special tryFuncImpl

tryFuncImpl :: [IR.IR Eval] -> Eval (IR.IR Eval)
tryFuncImpl (body : catches) = do
    runtime <- getRuntime
    result <- liftIO $ runEval (eval body) runtime
    case result of
        Right (val, newRuntime) -> do
            putRuntime newRuntime
            pure val
        Left (EvalError _ (RuntimeException sym payload)) -> do
            case findCatch sym catches of
                Just handler -> do
                    callable <- eval handler
                    case callable of
                        c | isCallable c -> case payload of
                            Just p -> apply c p -- Apply with single argument
                            Nothing -> apply c IR.Void -- Apply with void if no payload
                        _ -> throwError notCallableObject
                Nothing -> throwError $ RuntimeException sym payload
tryFuncImpl _ = throwError $ wrongArgumentType ["body", "catch*"]

findCatch :: Text -> [IR.IR Eval] -> Maybe (IR.IR Eval)
findCatch _ [] = Nothing
findCatch excSymbol (IR.List [IR.Symbol "catch", catchType, handler] : rest)
    | getSymbolText catchType == Just excSymbol = Just handler
    | otherwise = findCatch excSymbol rest
findCatch excSymbol (_ : rest) = findCatch excSymbol rest

getSymbolText :: IR.IR Eval -> Maybe Text
getSymbolText (IR.Symbol t) = Just t
getSymbolText _ = Nothing

File: dart/glue/lib/src/lib/builtin/try.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Try special form - exception handling with catch blocks
/// Mirrors Haskell Glue.Lib.Builtin.Try.tryFunc exactly
Eval<Ir> tryFunc(List<Ir> args) {
  if (args.isEmpty) {
    return throwError(wrongArgumentType(['body', 'catch*']));
  }

  final body = args[0];
  final catches = args.sublist(1);

  return getRuntime().flatMap((runtime) {
    return liftIO(runEval(eval(body), runtime)).flatMap((result) {
      return result.match(
        (error) {
          // Handle RuntimeException exactly like Haskell
          final runtimeExc = error.exception;
          final catchHandler = _findCatch(runtimeExc.symbol, catches);

          if (catchHandler != null) {
            return eval(catchHandler).flatMap((callable) {
              // Check if callable like Haskell does
              if (_isCallable(callable)) {
                // Use same payload handling as Haskell: maybe [] (: []) payload
                final List<Ir> payload = runtimeExc.value != null
                    ? [runtimeExc.value!]
                    : [];
                return apply(callable, payload);
              } else {
                return throwError(notCallableObject());
              }
            });
          } else {
            return throwError(runtimeExc);
          }
        },
        (success) {
          final (value, newRuntime) = success;
          return putRuntime(newRuntime).map((_) => value);
        },
      );
    });
  });
}

/// Find the appropriate catch handler for an exception name
Ir? _findCatch(String excSymbol, List<Ir> catches) {
  for (final catchClause in catches) {
    if (catchClause is IrList && catchClause.elements.length == 3) {
      final elements = catchClause.elements;
      if (elements[0] is IrSymbol &&
          (elements[0] as IrSymbol).value == 'catch') {
        final catchType = elements[1];
        final handler = elements[2];

        final catchTypeName = _getSymbolText(catchType);
        if (catchTypeName == excSymbol) {
          return handler;
        }
      }
    }
  }
  return null;
}

/// Extract symbol text from IR
String? _getSymbolText(Ir ir) {
  return switch (ir) {
    IrSymbol(value: final text) => text,
    _ => null,
  };
}

/// Check if an IR value is callable (closure or native function)
/// Mirrors Haskell isCallable
bool _isCallable(Ir ir) {
  return ir is IrClosure || (ir is IrNative && ir.value is NativeFunc);
}
